### PR TITLE
Fixes #2943,#11387 - Unset optional passwords

### PR DESCRIPTION
**Pull request by amirfefer**

### CODE REVIEW

**Comment by ShimShtein** [comment]
Wouldn't it be better to use pattern that resembles to that of "inherit" button in host's puppet fields?
My vision of this is the following:
By default the user will be presented with a disabled text field:
image
Because the field is disabled, it will not be passed as part of the `params` hash, and will not be updated in ActiveRecord.
If the user clicks the "set" button, the field will be enabled:
image
If the user will not type any text - the password will be set to blank (no password), if the user will type something - it will be the new password that will be set.

This way it will require less intervention on the server side (controller and model) and we already have part of the work done in #2230

**Comment by dLobatog** [comment]
@amirfefer The commit has a typo, also I think you can just shorten it to `unset optional passwords`. I've changed the pull request title.

The simpler we make these forms (which are already rather complicated) the better. I myself wouldn't know how to react if I saw a button "set" or "unset" next to a password field. What I think makes most sense and respects users' patterns:
- On _new_ forms, such as "New Host", "New Compute Resource", "New Host Group", "New Auth Source LDAP" (not sure if others need changing), the behaviour of setting the password works perfectly. **Don't make any change**. 
- On _edit_ forms: 
  - If there was **no password set**, we don't have to touch anything as **it works well**. 
  - If there was a password set, clearing it is the bug reported in Redmine. Do this: 
    - Substitute the Password text field by a button that says 'Reset' . 
    - Clicking on it makes an AJAX request to Foreman that clears the password for the object 
    - Re-enable the text field after clicking on 'Reset' in case the user wants to set it to something.

**Comment by ShimShtein** [comment]
@eLobato I totally agree about the "New *" forms, but for existing hosts, I think it would be better to avoid double roundtrip in case of setting the password to some different value. Basically it's the same design as I have proposed, just change the text to something more indicative: "Set password", "Override", "Change" or just "Unlock" can do he trick. Maybe it would be cleaner to use some glyph, like this: 
image

**Comment by dLobatog** [comment]
:+1: , 100% on board with the approach above.

**Comment by ohadlevy** [comment]
dont forget that in some cases, reset pw may means it defaults to hostgroup
defined password etc.

**Comment by ShimShtein** [comment]
@ohadlevy I think it's the same copy-on-create policy as with other fields. I would recommend giving up the feature of copying the password from the hostgroup for now at least. Although it's doable, it would create more UI elements on already heavy loaded form.

File: app/controllers/auth_source_ldaps_controller.rb [hunk 1]
```diff
@@ -22,8 +22,6 @@ def edit
   end
 
   def update
-    # remove from hash :account_password if blank?
-    params[:auth_source_ldap].except!(:account_password) if params[:auth_source_ldap][:account_password].blank?
```
Comment (amirfefer) [inline app/controllers/auth_source_ldaps_controller.rb:26]: Now the password field is disabled in edit forms by default, so no need to do that anymore.

File: app/assets/javascripts/application.js [hunk 2]
```diff
@@ -451,6 +451,17 @@ function setPowerState(item, status){
   $('[rel="twipsy"]').tooltip();
 }
 
+function disable_textfield(item) {
```
Comment (ohadlevy) [inline app/assets/javascripts/application.js]: name is misleading, afaiu it also enable disabled text fields? (or any input for that matter?)

File: app/controllers/auth_source_ldaps_controller.rb (hunk 1 above)
Comment (ohadlevy) [inline app/controllers/auth_source_ldaps_controller.rb:26]: is it always disabled?

File: app/helpers/layout_helper.rb [hunk 3]
```diff
@@ -50,15 +50,32 @@ def textarea_f(f, attr, options = {})
...
       options[:autocomplete]   ||= 'off'
       options[:placeholder]    ||= password_placeholder(f.object, attr)
+      options[:disabled] = true if unset_button
       addClass options, 'form-control'
-      f.password_field(attr, options) +
+      pass = f.password_field(attr, options) +
       '<span class="glyphicon glyphicon-warning-sign input-addon"
              title="'.html_safe + _('Caps lock ON') +
              '" style="display:none"></span>'.html_safe
+      if unset_button
```
Comment (ohadlevy) [inline app/helpers/layout_helper.rb:75]: nitpick: maybe extract to a separate method.

File: app/views/auth_source_ldaps/_form.html.erb [hunk 4]
```diff
@@ -21,7 +21,7 @@
     </div>
     <div class="tab-pane" id="account">
       <%= text_f f, :account, :help_inline =>_("Use this account to authenticate, <i>optional</i>").html_safe %>
-      <%= password_f f, :account_password, :help_inline => _("Use this account to authenticate, <i>optional</i>").html_safe %>
+      <%= password_f f, :account_password, :help_inline => _( "Use this account to authenticate, <i>optional</i>").html_safe, :unset => (true if action_name == "edit") %>
```
Comment (ohadlevy) [inline app/views/auth_source_ldaps/_form.html.erb]: please remove extra added white space before "Use

File: app/views/hostgroups/_form.html.erb [hunk 5]
```diff
@@ -61,7 +61,7 @@
 
     <div class="tab-pane" id="os">
       <%= render "common/os_selection/initial", :item=> @hostgroup %>
-      <%= password_f f, :root_pass, :help_inline => _("Password must be 8 characters or more")%>
+      <%= password_f f, :root_pass, :help_inline => _("Password must be 8 characters or more").html_safe, :unset => (true if action_name == "edit") %>
```
Comment (ohadlevy) [inline app/views/hostgroups/_form.html.erb]:
``` ruby
:unset => action_name == "edit"
```

File: app/controllers/auth_source_ldaps_controller.rb (hunk 1 above)
Comment (amirfefer) [inline app/controllers/auth_source_ldaps_controller.rb:26]: It looks like that:
pass
so unless the user click the change button, the password field is disabled

**Comment by dLobatog** [comment]
@amirfefer Tests are red due to a Katello failure, don't worry. It looks fine, just two more things before merging: 
- Please extract the code @ohadlevy pointed out to a separate method (thanks!)
- Hosts form may also contain a root_pass field, yet it's not protected as Hostgroups/Auth Source LDAP please make sure you use this helper in Hosts edit too. In fact this is issue [2943](http://projects.theforeman.org/issues/2943)

The name of the commit is a bit confusing in my opinion, if you could come up with something a bit more descriptive that'd be helpful.

Thanks again @amirfefer ! Almost there :smile:

File: app/controllers/hosts_controller.rb [hunk 6]
```diff
@@ -712,15 +711,4 @@ def merge_search_filter(filter)
   def offer_to_overwrite_conflicts
     @host.overwrite = "true" if @host.errors.any? and @host.errors.are_all_conflicts?
   end
-
-  # :root_pass and bmc :password should not update the Host and Nic passwords when blank
-  # This method removes them from the params hash to avoid any processing of these attributes.
-  def cleanup_passwords
```
Comment (amirfefer) [inline app/controllers/hosts_controller.rb:719]: not need anymore, the root_pass and the BMC password in interfaces_attributes are disabled in default

**Comment by amirfefer** [comment]
@eLobato I also added the option to  BMC password in host -interfaces_attributes additional to host root_pass

**Comment by domcleal** [comment]
This PR is in an unmergeable state, please rebase.

**Comment by dLobatog** [comment]
@amirfefer I think you forgot the field on Compute Resources, which is particularly harmful as it is now :smile: I've unset the password many times myself by mistake and wondered why I can't launch VMs.

:+1: otherwise, good to merge after that.

**Comment by dLobatog** [comment]
Merged as e63db5e7cf7da6f22334c7c8e7875ab47ec84d6f, thanks @amirfefer!

**Pull request closed without merging**